sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadIntrinsics.java

```diff
@@ -0,0 +1,847 @@
+        @Child
+        LLVMLoadNode read = null;
+
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex, Object attr) {
+            if (read == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                read = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createLoadNode(LLVMInteropType.ValueKind.I32);
+            }
+            // we can use the address of the native pointer here, bc a mutex
+            // must only work when using the original variable, not a copy
+            // so the address may never change
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
+            Object mutObj = LLVMLanguage.getLLVMContextReference().get().mutexStorage.get(mutexAddress);
+            int attrValue = 0;
+
+            if (!((LLVMPointer) attr).isNull()) {
+                attrValue = (int) read.executeWithTarget(attr);
+            }
+            Mutex.Type mutexType = Mutex.Type.DEFAULT_NORMAL;
+            if (attrValue == 1) {
+                mutexType = Mutex.Type.RECURSIVE;
+            } else if (attrValue == 2) {
+                mutexType = Mutex.Type.ERRORCHECK;
+            }
+            if (mutObj == null) {
+                LLVMLanguage.getLLVMContextReference().get().mutexStorage.put(mutexAddress, new Mutex(mutexType));
+            }
+            return 0;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadMutexLock extends LLVMBuiltin {
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex) {
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
+            Mutex mutexObj = (Mutex) LLVMLanguage.getLLVMContextReference().get().mutexStorage.get(mutexAddress);
+            if (mutexObj == null) {
+                // mutex is not initialized
+                // but it works anyway on most implementations
+                // so we will make it work here too, just using default type
+                // set the internLock counter to 1
+                mutexObj = new Mutex(Mutex.Type.DEFAULT_NORMAL);
+                LLVMLanguage.getLLVMContextReference().get().mutexStorage.put(mutexAddress, mutexObj);
+            }
+            // TODO: error code handling
+            return mutexObj.lock() ? 0 : 15;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadMutexTrylock extends LLVMBuiltin {
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object mutex) {
+            long mutexAddress = ((LLVMNativePointer) mutex).asNative();
+            Mutex mutexObj = (Mutex) LLVMLanguage.getLLVMContextReference().get().mutexStorage.get(mutexAddress);
+            if (mutexObj == null) {
+                // mutex is not initialized
+                // but it works anyway on most implementations
```

Comment:
What are those implementations? Does the spec say what should happen here and they adhere to it, or do they just happen to work in a case the spec does not mention?

Reply:
in the spec calling lock() before init() leads to unspecified behavior

on the implementation i got on my ubuntu 18.04 e. g. it works when you call lock() on a mutex_t that is zeroed out (calloc, global var, ...)
that's because a zeroed out mutex_t on most implementations (*) is a default unlocked mutex (what init() with default values and static initializing create)
i saw no reason to not also allow this on my implementation, because it's unspecified anyway, and people might expect that behavior

(*) i got that info from https://stackoverflow.com/questions/31663981/is-it-safe-to-call-pthread-mutex-lock-before-pthread-mutex-init

Reply:
Ok, that makes sense. Please remember corner cases like these for the thesis also.

Reply:
in such situations it is always a good idea to add a comment to your code why you made that decision, also citing the stackoverflow link as a reference